# What is the True Normal Human Body Temperature? 

#### Background

The mean normal body temperature was held to be 37$^{\circ}$C or 98.6$^{\circ}$F for more than 120 years since it was first conceptualized and reported by Carl Wunderlich in a famous 1868 book. But, is this value statistically correct?

<h3>Exercises</h3>

<p>In this exercise, you will analyze a dataset of human body temperatures and employ the concepts of hypothesis testing, confidence intervals, and statistical significance.</p>

<p>Answer the following questions <b>in this notebook below and submit to your Github account</b>.</p> 

<ol>
<li>  Is the distribution of body temperatures normal? 
    <ul>
    <li> Although this is not a requirement for the Central Limit Theorem to hold (read the introduction on Wikipedia's page about the CLT carefully: https://en.wikipedia.org/wiki/Central_limit_theorem), it gives us some peace of mind that the population may also be normally distributed if we assume that this sample is representative of the population.
    <li> Think about the way you're going to check for the normality of the distribution. Graphical methods are usually used first, but there are also other ways: https://en.wikipedia.org/wiki/Normality_test
    </ul>
<li>  Is the sample size large? Are the observations independent?
    <ul>
    <li> Remember that this is a condition for the Central Limit Theorem, and hence the statistical tests we are using, to apply.
    </ul>
<li>  Is the true population mean really 98.6 degrees F?
    <ul>
    <li> First, try a bootstrap hypothesis test.
    <li> Now, let's try frequentist statistical testing. Would you use a one-sample or two-sample test? Why?
    <li> In this situation, is it appropriate to use the $t$ or $z$ statistic? 
    <li> Now try using the other test. How is the result be different? Why?
    </ul>
<li>  Draw a small sample of size 10 from the data and repeat both frequentist tests. 
    <ul>
    <li> Which one is the correct one to use? 
    <li> What do you notice? What does this tell you about the difference in application of the $t$ and $z$ statistic?
    </ul>
<li>  At what temperature should we consider someone's temperature to be "abnormal"?
    <ul>
    <li> As in the previous example, try calculating everything using the boostrap approach, as well as the frequentist approach.
    <li> Start by computing the margin of error and confidence interval. When calculating the confidence interval, keep in mind that you should use the appropriate formula for one draw, and not N draws.
    </ul>
<li>  Is there a significant difference between males and females in normal temperature?
    <ul>
    <li> What testing approach did you use and why?
    <li> Write a story with your conclusion in the context of the original problem.
    </ul>
</ol>

You can include written notes in notebook cells using Markdown: 
   - In the control panel at the top, choose Cell > Cell Type > Markdown
   - Markdown syntax: http://nestacms.com/docs/creating-content/markdown-cheat-sheet

#### Resources

+ Information and data sources: http://www.amstat.org/publications/jse/datasets/normtemp.txt, http://www.amstat.org/publications/jse/jse_data_archive.htm
+ Markdown syntax: http://nestacms.com/docs/creating-content/markdown-cheat-sheet

****

In [1]:
import pandas as pd
import numpy as np
from scipy import stats
import plotly as py
import plotly.graph_objs as go
import plotly.figure_factory as ff

py.offline.init_notebook_mode(connected=True)
df = pd.read_csv('data/human_body_temperature.csv')

***
<div class="span5 alert alert-success">
    <h4>Question 1: Is the distribution of body temperatures normal?</h4>
</div>

In [2]:
df.head()

temperature gender  heart_rate
0         99.3      F        68.0
1         98.4      F        81.0
2         97.8      M        73.0
3         99.2      F        66.0
4         98.0      F        73.0

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 130 entries, 0 to 129
Data columns (total 3 columns):
temperature    130 non-null float64
gender         130 non-null object
heart_rate     130 non-null float64
dtypes: float64(2), object(1)
memory usage: 3.1+ KB


In [4]:
unique_temps = df.temperature.unique()
unique_temps.sort()
print(unique_temps)
print(len(unique_temps))

[ 96.3  96.4  96.7  96.8  96.9  97.   97.1  97.2  97.3  97.4  97.5  97.6
  97.7  97.8  97.9  98.   98.1  98.2  98.3  98.4  98.5  98.6  98.7  98.8
  98.9  99.   99.1  99.2  99.3  99.4  99.5  99.9 100.  100.8]
34


In [5]:
def ecdf(data):
    """Computes the Empirical Cumulative Distribution Function of a 1-dimensional numerical array"""
    n = len(data)
    x = np.sort(data)
    y = np.arange(1, n + 1) / n
    return x, y

In [6]:
x_temp, y_temp = ecdf(df.temperature)

sample_mean = np.mean(df.temperature)
sample_stdv = np.std(df.temperature)
samps_nrm = np.random.normal(sample_mean, sample_stdv, 10000)
x_nrml, y_nrml = ecdf(samps_nrm)
x_mean, y_mean = x_mean, y_mean = [[sample_mean, sample_mean], [-0.05, 1.05]]


trace0 = go.Histogram(name="Actual Samples", x=x_temp, histnorm="probability", nbinsx=100)
trace1 = go.Scatter(name="Actual Samples", x=x_temp, y=y_temp, mode="markers", marker=dict(color="rgba(31,119,180,1)"))
trace2 = go.Scatter(name="Normal Distribution", x=x_nrml, y=y_nrml, mode="line")
trace3 = go.Scatter(x=x_mean, y=y_mean, mode="lines", line=dict(color="red"), name="Actual Sample Mean")

fig = py.tools.make_subplots(rows=2, cols=1, shared_xaxes=True, vertical_spacing=0.01)
fig.append_trace(trace0, 1, 1)
fig.append_trace(trace1, 2, 1)
fig.append_trace(trace2, 2, 1)
fig.append_trace(trace3, 2, 1)

fig["layout"].update(
    plot_bgcolor="rgb(247,247,247)",
    legend=dict(font=dict(family="serif", size=12)),
    height=600,
    title="<b>Human Body Temperature Sample Distribution</b>",
    titlefont=dict(family="serif", size=24),
    yaxis1=dict(title="<b>PDF</b>", titlefont=dict(family="serif", size=14),
               tickfont=dict(family="serif", size=14)), 
    yaxis2=dict(title="<b>ECDF</b>", titlefont=dict(family="serif", size=14), dtick=0.2,
               tickfont=dict(family="serif", size=14), range=[-0.05, 1.05]),
    xaxis1=dict(title="<b>Human Body Temperature, {0}F</b>".format(u'\xb0'), dtick=0.5,
               titlefont=dict(family="serif", size=14)))

py.offline.iplot(fig)

This is the format of your plot grid:
[ (1,1) x1,y1 ]
[ (2,1) x1,y2 ]



With some quick, graphical exploratory data analysis (EDA) the human body temperature sample distribution appears to be normally distributed.

***
<div class="span5 alert alert-success">
    <h4>Question 2: Is the sample size large? Are the observations independent?</h4>
</div>

Yes, the sample size can be considered large enough (n >= 30) for statistical inference because there are 130 samples (n = 130). Yes, the samples can be considered as independent observations without replacement because the sample size is well below 10% of the population being considered (all humans).

***
<div class="span5 alert alert-success">
    <h4>Question3: Is the true population mean really 98.6 degrees F?</h4>
    <ul>
        <li> First, try a bootstrap hypothesis test.
        <li> Now, let's try frequentist statistical testing. Would you use a one-sample or two-sample test? Why?
        <li> In this situation, is it appropriate to use the $t$ or $z$ statistic? 
        <li> Now try using the other test. How is the result be different? Why?
    </ul>
</div>

We need to start this statistical inference analysis by setting up some hypothesis testing parameters. The parameters I used are as follows:  
  
**Null Hypothesis:** The Population Mean Human Body Temperature is equal to 98.6 degrees F  
**Alternative Hypothesis:** The Population Mean Human Body Temperature is NOT equal to 98.6 degrees F  
**Confidence Interval:** 95%  
**Signifigence Level:** 5%  

In [7]:
def draw_bs_replicates(data, func, size=1):
    """Computes the bootstrap replicate of a 1-dimensional numerical array"""
    bs_replicates = np.empty(size)
    
    for i in range(size):
        bs_sample = np.random.choice(data, size=len(data))
        bs_replicates[i] = func(bs_sample)
        
    return bs_replicates

In [8]:
shifted_temp = df.temperature + 98.6 - sample_mean
shifted_bs_means = draw_bs_replicates(shifted_temp, np.mean, 1000)
x_bsrp, y_bsrp = ecdf(shifted_bs_means)

trace0 = go.Histogram(name="Shifted Bootstrap Means", x=x_bsrp, histnorm="probability", nbinsx=100)
trace1 = go.Scatter(x=x_bsrp, y=y_bsrp, mode="markers", name="Shifted Bootstrap Means",
                    marker=dict(color="rgba(31, 119, 180, 1)"))
trace2 = go.Scatter(x=x_mean, y=y_mean, mode="lines", line=dict(color="red"), name="Actual Sample Mean")

fig = py.tools.make_subplots(rows=2, cols=1, shared_xaxes=True, vertical_spacing=0.01)
fig.append_trace(trace0, 1, 1)
fig.append_trace(trace1, 2, 1)
fig.append_trace(trace2, 2, 1)

fig["layout"].update(
    plot_bgcolor="rgb(247,247,247)",
    legend=dict(font=dict(family="serif", size=12)),
    height=600,
    title="<b>Shifted Bootstrap Human Body Temperature Means Compared to Actual Sample Mean</b>",
    titlefont=dict(family="serif", size=24),
    yaxis1=dict(title="<b>PDF</b>", titlefont=dict(family="serif", size=14),
               tickfont=dict(family="serif", size=14)), 
    yaxis2=dict(title="<b>ECDF</b>", titlefont=dict(family="serif", size=14), dtick=0.2,
               tickfont=dict(family="serif", size=14), range=[-0.05, 1.05]),
    xaxis1=dict(title="<b>Human Body Temperature, {0}F</b>".format(u'\xb0'), dtick=0.1,
               titlefont=dict(family="serif", size=14), range=[98, 99]))

py.offline.iplot(fig)

bs_p = np.sum(shifted_bs_means <= np.mean(df.temperature)) / len(shifted_bs_means) * 2 # two-sided p-value
print("The probability that the Population Mean Human Body Temperature is equal to 98.6{0}F is "
      .format(u'\xb0') + str(bs_p) + " when considering the provided sample data.")

This is the format of your plot grid:
[ (1,1) x1,y1 ]
[ (2,1) x1,y2 ]



The probability that the Population Mean Human Body Temperature is equal to 98.6°F is 0.0 when considering the provided sample data.


The **bootstrap replication hypothesis test** above suggests that we reject the null hypothesis and support the alternative hypothesis that the mean human body temperature is NOT 98.6 degrees F when considering the provided sample data.

I have used a one-sample inference test for the **frequentist statistical analysis**  because we have been provided the historical mean human body temperature (98.6 degrees F) which represents the population mean (mu) and a sample set of human body temperature measurements where we can extract the sample mean (x-bar) and the sample standard deviation (Sx).  

In the situation of analyzing statictical inference about the population mean, I'm choosing to use the t-statistic because the standard deviation of the population (sigma) is unknown.

In [9]:
mu = 98.6
t = (sample_mean - mu) / (sample_stdv / (len(df.temperature))**0.5)
t_p = stats.t.sf(np.abs(t), len(df.temperature) - 1) * 2 # two-sided p-value

print("The null hypothesis population mean is " + str(mu) + "{0}F.".format(u'\xb0'))
print("The sample mean is " + str(sample_mean) + "{0}F.".format(u'\xb0'))
print("The sample standard deviation is " + str(sample_stdv) + "{0}F.".format(u'\xb0'))
print("The t-statistic for this inference is " + str(t) + ".")
print("The degree of freedom for this sample set is " + str(len(df.temperature) - 1) + ".")
print("The probability that the Population Mean Human Body Temperature is equal to 98.6{0}F is "
      .format(u'\xb0') + str(t_p) + " when considering the provided sample data.")

The null hypothesis population mean is 98.6°F.
The sample mean is 98.24923076923078°F.
The sample standard deviation is 0.7303577789050377°F.
The t-statistic for this inference is -5.475925202078116.
The degree of freedom for this sample set is 129.
The probability that the Population Mean Human Body Temperature is equal to 98.6°F is 2.1887464624095175e-07 when considering the provided sample data.


The **frequentist statistical hypothesis testing** above also suggests that we reject the null hypothesis and support the alternative hypothesis that the mean human body temperature is NOT 98.6 degrees F when considering the provided sample data.

In [10]:
z = t
z_p = stats.norm.sf(abs(z)) * 2 # two-sided p-value

print("The probability that the Population Mean Human Body Temperature is equal to 98.6{0}F is "
      .format(u'\xb0') + str(z_p) + " when considering the provided sample data.")

The probability that the Population Mean Human Body Temperature is equal to 98.6°F is 4.3523151658821886e-08 when considering the provided sample data.


If we approximate the z-statistic equal to the t-statistic and then evaluate the probability, then the probability decreases as the z-probability distribution (the standard normal probability distribution) has a tighter probability distribuiton around the mean than the t-probability distribution.

***
<div class="span5 alert alert-success">
    <h4>Question 4: Draw a small sample of size 10 from the data and repeat both frequentist tests.</h4>
    <ul>
        <li> Which one is the correct one to use? 
        <li> What do you notice? What does this tell you about the difference in application of the $t$ and $z$ statistic?
    </ul>
</div>

In [11]:
small_sample = np.random.choice(df.temperature, size=10)
small_sample_mean = np.mean(small_sample)
small_sample_stdv = np.std(small_sample)
small_t = (small_sample_mean - mu) / (small_sample_stdv / (len(small_sample))**0.5)
small_t_p = stats.t.sf(np.abs(small_t), len(small_sample) - 1) * 2 # two-sided p-value

print("The null hypothesis population mean is " + str(mu) + "{0}F.".format(u'\xb0'))
print("The small sample mean is " + str(small_sample_mean) + "{0}F.".format(u'\xb0'))
print("The small sample standard deviation is " + str(small_sample_stdv) + "{0}F.".format(u'\xb0'))
print("The t-statistic for this inference is " + str(small_t) + ".")
print("The degree of freedom for this small sample set is " + str(len(small_sample) - 1) + ".")
print("The probability that the Population Mean Human Body Temperature is equal to 98.6{0}F is "
      .format(u'\xb0') + str(small_t_p) + " when considering a small sample of the sample data utilizing the t-statistic.")
print()

small_z = small_t
small_z_p = stats.norm.sf(abs(small_z)) * 2 # two-sided p-value

print("The probability that the Population Mean Human Body Temperature is equal to 98.6{0}F is "
      .format(u'\xb0') + str(small_z_p) + " when considering a small sample of the sample data utilizing the z-statistic.")

The null hypothesis population mean is 98.6°F.
The small sample mean is 98.55999999999999°F.
The small sample standard deviation is 0.5480875842417892°F.
The t-statistic for this inference is -0.23078630139330675.
The degree of freedom for this small sample set is 9.
The probability that the Population Mean Human Body Temperature is equal to 98.6°F is 0.8226416779487429 when considering a small sample of the sample data utilizing the t-statistic.

The probability that the Population Mean Human Body Temperature is equal to 98.6°F is 0.8174808238892333 when considering a small sample of the sample data utilizing the z-statistic.


The correct statistic to use is the t-static because this small sample set is less than 30 (n < 30) and we don't know the standard deviation of the population (sigma). The z-statistic is going to result in a smaller probability which could falsely reject the null hypothesis.

***
<div class="span5 alert alert-success">
    <h4>Question 5: At what temperature should we consider someone's temperature to be "abnormal"?</h4>
    <ul>
        <li> As in the previous example, try calculating everything using the boostrap approach, as well as the frequentist approach.
        <li> Start by computing the margin of error and confidence interval. When calculating the confidence interval, keep in mind that you should use the appropriate formula for one draw, and not N draws.
        </ul>
</div>

In [12]:
bs_means = draw_bs_replicates(df.temperature, np.mean, 1000)
bs_conf_int = np.percentile(bs_means, [2.5, 97.5])

print("The bootstrap replicated mean 95% confidence interval is " + str(bs_conf_int))

t_95_129 = stats.t.ppf(1 - 0.025, len(df.temperature) - 1) # two-tail for 95% confidence interval
conf_int = np.array([sample_mean - (t_95_129 * sample_stdv / (len(df.temperature) ** 0.5)),
                     sample_mean + (t_95_129 * sample_stdv / (len(df.temperature) ** 0.5))])

print("The frequentist mean 95% confidence interval utilizing the t-statistic is " + str(conf_int))

The bootstrap replicated mean 95% confidence interval is [98.13228846 98.37546154]
The frequentist mean 95% confidence interval utilizing the t-statistic is [98.12249319 98.37596835]


According to the resulting 95% confidence interval(s) obtained above from the provided sample data, an "abnormal" human body temperature could be any temperature below 98.1 and above 98.4 degrees F.

***
<div class="span5 alert alert-success">
    <h4>Question 6: Is there a significant difference between males and females in normal temperature?</h4>
    <ul>
        <li> What testing approach did you use and why?
        <li> Write a story with your conclusion in the context of the original problem.
    </ul>
</div>

We need to start this statistical inference analysis by setting up some hypothesis testing parameters. The parameters I used are as follows:  
  
**Null Hypothesis:** The difference between Male and Female Population Mean Human Body Temperatures is equal to 0 degrees F  
**Alternative Hypothesis:** The diffence between Male and Female Population Mean Human Body Temperatures is NOT equal to 0 degrees F  
**Confidence Interval:** 95%  
**Signifigence Level:** 5%  

In [13]:
male_samples = df.temperature[df.gender == "M"]
female_samples = df.temperature[df.gender == "F"]
male_mean = np.mean(male_samples)
female_mean = np.mean(female_samples)
x_mmean, y_mmean = np.array([[male_mean, male_mean], [-0.05, 1.05]])
x_fmean, y_fmean = np.array([[female_mean, female_mean], [-0.05, 1.05]])
x_male, y_male = ecdf(male_samples)
x_female, y_female = ecdf(female_samples)

trace0 = go.Histogram(name="Male Samples", x=male_samples, histnorm="probability", 
                      opacity=0.75, nbinsx=50)
trace1 = go.Histogram(name="Female Samples", x=female_samples, histnorm="probability", 
                      opacity=0.75, nbinsx=50)
trace2 = go.Scatter(name="Male Samples", x=x_male, y=y_male, mode="markers", 
                    marker=dict(color="rgba(87, 153, 199, 1)"))
trace3 = go.Scatter(name="Female Samples", x=x_female, y=y_female, mode="markers",
                    marker=dict(color="rgba(255, 159, 74, 1)"))
trace4 = go.Scatter(name="Male Sample Mean", x=x_mmean, y=y_mmean, mode="lines",
                    line=dict(color="rgba(87, 153, 199, 1)"))
trace5 = go.Scatter(name="Female Sample Mean", x=x_fmean, y=y_fmean, mode="lines",
                    line=dict(color="rgba(255, 159, 74, 1)"))

fig = py.tools.make_subplots(rows=2, cols=1, shared_xaxes=True, vertical_spacing=0.01)
fig.append_trace(trace0, 1, 1)
fig.append_trace(trace1, 1, 1)
fig.append_trace(trace2, 2, 1)
fig.append_trace(trace3, 2, 1)
fig.append_trace(trace4, 2, 1)
fig.append_trace(trace5, 2, 1)

fig["layout"].update(
    plot_bgcolor="rgb(247,247,247)",
    legend=dict(font=dict(family="serif", size=12)),
    height=600,
    title="<b>Human Body Temperature Sample Distribution by Gender</b>",
    titlefont=dict(family="serif", size=24),
    yaxis1=dict(title="<b>PDF</b>", titlefont=dict(family="serif", size=14),
               tickfont=dict(family="serif", size=14)), 
    yaxis2=dict(title="<b>ECDF</b>", titlefont=dict(family="serif", size=14), dtick=0.2,
               tickfont=dict(family="serif", size=14), range=[-0.05, 1.05]),
    xaxis1=dict(title="<b>Human Body Temperature, {0}F</b>".format(u'\xb0'), dtick=0.5,
               titlefont=dict(family="serif", size=14)))

py.offline.iplot(fig)

pop_mean_diff = 0
sample_mean_diff = abs(male_mean - female_mean)
sample_mean_diff_stdv = (np.var(male_samples) / len(male_samples) 
                         + np.var(female_samples) / len(female_samples)) ** 0.5
t = (sample_mean_diff - pop_mean_diff) / (sample_mean_diff_stdv / (len(male_samples) 
                                                                   + len(female_samples)) ** 0.5)
t_p = stats.t.sf(np.abs(t), (len(male_samples) + len(female_samples)) - 1) * 2 # two-sided p-value

print("The null hypothesis difference between the male and female population mean is " 
      + str(pop_mean_diff) + "{0}F.".format(u'\xb0'))
print("The difference between male and female sample means is " + str(sample_mean_diff) 
      + "{0}F.".format(u'\xb0'))
print("The standard deviation of the difference between male and female sample mean is " 
      + str(sample_mean_diff_stdv) + "{0}F.".format(u'\xb0'))
print("The t-statistic for this inference is " + str(t) + ".")
print("The degree of freedom for this sample set is " + str((len(male_samples) + len(female_samples)) - 1) + ".")
print("The probability that the difference between Male and Female Population Mean Human Body Temperature is equal to 0{0}F is "
      .format(u'\xb0') + str(t_p) + " when considering the provided sample data.")

This is the format of your plot grid:
[ (1,1) x1,y1 ]
[ (2,1) x1,y2 ]



The null hypothesis difference between the male and female population mean is 0°F.
The difference between male and female sample means is 0.289230769230727°F.
The standard deviation of the difference between male and female sample mean is 0.12557668521231213°F.
The t-statistic for this inference is 26.260751723287424.
The degree of freedom for this sample set is 129.
The probability that the difference between Male and Female Population Mean Human Body Temperature is equal to 0°F is 1.3225249126772841e-53 when considering the provided sample data.
